added xml docs to concepts

Mohamed-Dhouioui opened the pull request:
I've created an xml-files page under concepts in the docs to showcase the XML format. I'm not sure this is the best way to do it, so I'd appreciate all your feedback.  I'll also add the information from the docstrings into the Openmm serializing notebook tutorial to make life easier for anyone using it.

rhoslynroebuck commented on source/concepts/xml-files.rst:
```diff
@@ -0,0 +1,57 @@
+The XML format for serializing OpenMM simulations consists of a root element ``<OpenMMSimulation>`` containing three main components:
+
+1. **Starting Structure**: Enclosed in either ``<pdbx>`` or ``<pdb>`` tags
+2. **OpenMM Serialized System**: Enclosed in ``<System>`` tags
```
```suggestion
2. **OpenMM Serialized System**: Enclosed in the ``<System>`` tag
```

rhoslynroebuck replied:
I think this should be singular since there can only be one System

rhoslynroebuck commented on source/concepts/xml-files.rst:
```diff
@@ -0,0 +1,57 @@
+
+1. **Starting Structure**: Enclosed in either ``<pdbx>`` or ``<pdb>`` tags
+2. **OpenMM Serialized System**: Enclosed in ``<System>`` tags
+3. **OpenMM Serialized Integrator**: Enclosed in ``<Integrator>`` tags
```
```suggestion
3. **OpenMM Serialized Integrator**: Enclosed in the ``<Integrator>`` tag
```

rhoslynroebuck replied:
I think this should be singular since there can only be one Integrator

rhoslynroebuck commented on source/concepts/xml-files.rst:
```diff
@@ -0,0 +1,57 @@
+2. **OpenMM Serialized System**: Enclosed in ``<System>`` tags
+3. **OpenMM Serialized Integrator**: Enclosed in ``<Integrator>`` tags
+
+An optional fourth component, the OpenMM serialized state, may also be included.
```
Can you describe more about what this is? What type of object is it and what is the tag?

Mohamed-Dhouioui replied:
Done in this commit [84ca9a9](https://github.com/IRL2/nanover-docs/pull/115/commits/84ca9a93f9ad2f5f4ace78b0b3f69b84b985e283)

rhoslynroebuck commented on source/concepts/xml-files.rst:
```diff
@@ -0,0 +1,57 @@
+Usage
+-----
+
+This XML format serves two primary purposes:
```
```suggestion
The `narupa.openmm.serializer` module provides the following functions to convert to/from a serialized OpenMM simulation:
```

Mohamed-Dhouioui replied:
I think there is a typo there, it should be: `nanover,openmm.serializer`

rhoslynroebuck commented on source/concepts/xml-files.rst:
```diff
@@ -0,0 +1,57 @@
+1. **Serialization**: Converting an existing OpenMM simulation instance into an XML file
+2. **Deserialization**: Creating an OpenMM simulation instance from an XML file
+
+The format allows for efficient storage and exchange of OpenMM simulation data, including the starting structure, system configuration, and integrator settings.
```
Not sure we need this line? Or maybe it further up the page.

Mohamed-Dhouioui replied:
I agree, and moved it up.

rhoslynroebuck commented on source/concepts/xml-files.rst:
```diff
@@ -0,0 +1,57 @@
+XML Files
```
I think the title should reflect the idea that this is the XML format for OpenMM simulations specifically.

Mohamed-Dhouioui replied:
Done in this commit [53f559a](https://github.com/IRL2/nanover-docs/pull/115/commits/53f559a4555a4359a66443667809392abf6f96e3).

rhoslynroebuck reviewed (changes requested):
Nice start on this page!

I've left a few specific comments, and then here are a couple of more general things:

- I think, since this refers specifically to the OpenMM package, maybe it should be put somewhere else, maybe in the [OpenMM tutorials page](https://irl2.github.io/nanover-docs/tutorials/openmm.html)?
- Also, it would be good to include a link to the specific module that this info is referring to.

hjstroud commented on source/concepts/xml-files.rst:
```diff
@@ -0,0 +1,88 @@
+
+    **Forces:** Various force components are represented by specific tags:
+
+        - ``<HarmonicBondForce>``: For bond stretching
```
These are some examples of the forces that one could include in an input file, but it is not an exhaustive list. Perhaps it would be better to refer to the types of forces that [OpenMM offers](http://docs.openmm.org/latest/userguide/theory/02_standard_forces.html#standard-forces) and link to the specific page?

Mohamed-Dhouioui replied:
Solved in commit [8e32555](https://github.com/IRL2/nanover-docs/pull/115/commits/8e325552911b99f8b0b9deb777237b6f1ec7784d)

hjstroud commented on source/concepts/xml-files.rst:
```diff
@@ -0,0 +1,88 @@
+3. **OpenMM Serialized Integrator**:
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``<Integrator>`` tag contains parameters specific to the integration method used, such as step size and temperature for a Langevin integrator:
```
```suggestion
The ``<Integrator>`` tag contains parameters that specify the integration method to be used to simulate dynamics, such as the type of integrator, simulation time step and temperature:
```

hjstroud replied:
It may also be worth linking to the [integrators page](http://docs.openmm.org/latest/userguide/theory/04_integrators.html) of the OpenMM documentation

Mohamed-Dhouioui replied:
Soved in commit [8e32555](https://github.com/IRL2/nanover-docs/pull/115/commits/8e325552911b99f8b0b9deb777237b6f1ec7784d)

hjstroud reviewed (commented):
This is definitely a good start for making this information more visible. I have a couple of comments:

1. At the moment this information doesn't reflect all of the possible functionality. It might be worth taking a look at the comments in [openmm/serializer.py](https://github.com/IRL2/nanover-protocol/blob/1daeb9b517fcec21b200f6057f1a21c6483eb4e2/python-libraries/nanover-openmm/src/nanover/openmm/serializer.py#L1). One of the key bits of information missing from this PR is that you can optionally save the state of the system, which will be a very important thing for researchers who wish to save their equilibrated systems in the state that they were in after equilibration. Perhaps incorporating some of the information in this documentation, and pointing towards where it can be found (i.e. [here](https://irl2.github.io/nanover-docs/python/nanover.openmm.serializer.html)) would be a good idea, as this documentation should be in line with the functionality of the code.
2. Currently this file is placed in the `Concepts` section, but it is actually only relevant to OpenMM simulations. I would suggest that either this information be moved to the OpenMM tutorials section (to keep all of the information about OpenMM simulations together) _or_ that this file should be renamed to something like `input_files.rst`, and the information about the OpenMM input files should be a subsection of this.

I'd be happy to talk through any of this if it doesn't make sense/discuss appropriate solutions if that would be of any use!

Mohamed-Dhouioui commented:
Thanks for the feedback Harry & Rhos. It is very helpful and I think this page looks way better now. Please have a look at these changes:

- Added information about forces and integrators and links to their corresponding documentation in openmm.
- Added usage overview and explanation of the serializer module with some examples.
- Moved the xml-files.rst file into the openmm tutorials page.

rhoslynroebuck commented on source/tutorials/openmm.rst:
```diff
@@ -17,4 +23,128 @@ folder of the GitHub repository. It contains:
\ No newline at end of file
+  different atom selections for a protein-ligand system.
+
+XML Files
```
I think maybe this heading ("XML Files") could be refactored to something that makes it clear that these are NanoVer simulation files of an OpenMM simulation, rather than being a general XML file. It could also be good to add a sentence below this heading to explain this.

hjstroud replied:
Perhaps something like "NanoVer OpenMM input files"? XML is just a general nested file structure, not specific to NanoVer or OpenMM, and the important thing is that the input files described in this section are specific to OpenMM simulations performed using NanoVer

Mohamed-Dhouioui replied:
The XML format we are using does not seem to be different from the one OpenMM defines in their documentation. As far as I know, it is not proprietary to NanoVer (The same Serializer Class exists in Openmm with the same XML structure [here](http://docs.openmm.org/latest/api-python/generated/openmm.openmm.XmlSerializer.html?highlight=serializer#openmm.openmm.XmlSerializer) )

rhoslynroebuck replied:
So there is a slight difference between the OpenMM format and the NanoVer format. NanoVer input files for OpenMM simulations contain a pdb and a number of xml serialized objects. Although the serialized objects are identical to those used by OpenMM (they are actually created used OpenMM), the overall contents and structure of the files is specific to NanoVer and cannot be read by OpenMM on it's own. Explaining this could prevent users misunderstanding the NanoVer input file format.

Mohamed-Dhouioui replied:
Thanks for clearing this out @rhoslynroebuck. I agree that specifying the overall contents and structure of the XML files in the docs would prevent users from misinterpreting or misusing them.

Mohamed-Dhouioui replied:
Resolved in commit [f62ca6e](https://github.com/IRL2/nanover-docs/pull/115/commits/f62ca6e5fd4d7a15e0d7843f064081e5d054ccf2)

rhoslynroebuck commented on source/tutorials/openmm.rst:
```diff
@@ -17,4 +23,128 @@ folder of the GitHub repository. It contains:
+An optional fourth component, the OpenMM serialized state, may also be included.
+The format allows for efficient storage and exchange of OpenMM simulation data, including the starting structure, system configuration, and integrator settings.
+
+XML Structure
```
Each word starts with an upper case in the sub headings on this page, which is inconsistent with the rest of the docs. Would be good to change these so only the first letter starts with an upper case.

rhoslynroebuck commented on source/tutorials/openmm.rst:
```diff
@@ -17,4 +23,128 @@ folder of the GitHub repository. It contains:
+
+Within this tag, you'll find:
+
+    **Particles:** They are represented with ``<Particle>`` tags, each containing:
```
This line is all in bold (default for a nested list in Sphinx). Adding an extra line break stops this (https://stackoverflow.com/questions/42297839/parent-element-of-nested-list-is-bold)

rhoslynroebuck commented on source/tutorials/openmm.rst:
```diff
@@ -17,4 +23,128 @@ folder of the GitHub repository. It contains:
+Serializing a Simulation
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+To save a simulation to XML::
```
Can you make this into a code block?

rhoslynroebuck commented on source/tutorials/openmm.rst:
```diff
@@ -17,4 +23,128 @@ folder of the GitHub repository. It contains:
+
+To load a simulation from XML::
+
+    with open("sim.xml", "r") as f:
```
Can you make this into a code block?

rhoslynroebuck commented on source/tutorials/openmm.rst:
```diff
@@ -17,4 +23,128 @@ folder of the GitHub repository. It contains:
+
+For example::
+
+    simulation = nanover.openmm.serializer.deserialize_simulation(
```
Can you make this into a code block?

rhoslynroebuck commented on source/tutorials/openmm.rst:
```diff
@@ -17,4 +23,128 @@ folder of the GitHub repository. It contains:
+3. **OpenMM Serialized Integrator**: Enclosed in the ``<Integrator>`` tag
+
+An optional fourth component, the OpenMM serialized state, may also be included.
+The format allows for efficient storage and exchange of OpenMM simulation data, including the starting structure, system configuration, and integrator settings.
```
Can you link to some documentation about what an "OpenMM serialized state" is?

Mohamed-Dhouioui replied:
I read about it in the openmm documentation [here](http://docs.openmm.org/7.7.0/userguide/theory/05_other_features.html?highlight=serialized%20state#xmlserializer) but nothing more exists about it. I wanted to keep the information there in case someone know what it is and needs it.

hjstroud replied:
I think that OpenMM just has a method of serializing the objects it uses (be is a system, integrator, state etc.), and the label "OpenMM serialized" just indicates that it's been created using the functionality of OpenMM. So rather than explaining what an OpenMM serialized state is, it may be better to put in a short sentence to say that the fields prefixed by "OpenMM serialized" are created using the OpenMM inbuilt functionality to do so.

Mohamed-Dhouioui replied:
Solved in this commit [0043457](https://github.com/IRL2/nanover-docs/pull/115/commits/0043457c7c21028f743a19611dc6a729dd6a82c8)

rhoslynroebuck commented on source/tutorials/openmm.rst:
```diff
@@ -17,4 +23,128 @@ folder of the GitHub repository. It contains:
+Deserializing a Simulation
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To load a simulation from XML::
```
I think it would be good to make it clear that this is loading a simulation from a NanoVer OpenMM XML file, not just a general XML file. (Same comment for below)

Mohamed-Dhouioui replied:
resolved in commit [f62ca6e](https://github.com/IRL2/nanover-docs/pull/115/commits/f62ca6e5fd4d7a15e0d7843f064081e5d054ccf2)

rhoslynroebuck reviewed (changes requested):
Just a few formatting issues and a couple of clarifications

Mohamed-Dhouioui commented:
Thank you guys, these are the changes that I have done:

1. Renaming and clarifying titles and descriptions for better readability and specificity:
   - Changed "XML Files" to "Openmm XML Files": Because as far as I know we are using the same XML format as OpenMM specifies in their documentation.
   - Updated the description of the XML format to specify that it is used for manipulating NanoVer simulations from OpenMM serialized XML files.
   - Changed "XML Structure" to "XML structure".
   - Changed "Components Description" to "Components description".
   - Changed "Starting Structure" to "Starting structure".
   - Changed "OpenMM Serialized System" to "OpenMM serialized system".
   - Changed "OpenMM Serialized Integrator" to "OpenMM serialized integrator".
   - Changed "Serializing a Simulation" to "Serializing a simulation".
   - Changed "Deserializing a Simulation" to "Deserializing a simulation".

2. Added a brief note describing the XML format's specific use in NanoVer simulations.

3. Minor formatting adjustments, such as adding a blank line after the "Particles" tag description.

hjstroud commented on source/tutorials/openmm.rst:
```diff
@@ -17,4 +23,128 @@ folder of the GitHub repository. It contains:
+        <Integrator ...>
+            <!-- XML content of the OpenMM serialized integrator -->
+        </Integrator>
+    </OpenMMSimulation>
```
May be worth keeping an eye on IRL2/nanover-protocol#285, as this will add the `State` field too

Mohamed-Dhouioui replied:
Done in this commit [0043457](https://github.com/IRL2/nanover-docs/pull/115/commits/0043457c7c21028f743a19611dc6a729dd6a82c8).

hjstroud commented on source/tutorials/openmm.rst:
```diff
@@ -17,4 +23,129 @@ folder of the GitHub repository. It contains:
+        xml_string,
+        platform_name="CUDA"
+    )
+
```
May be worth linking to an OpenMM tutorial that demonstrates how to write OpenMM input files (e.g. the polyalanine notebook in the OpenMM examples directory)

Mohamed-Dhouioui replied:
Is [this](https://github.com/openmm/openmm-cookbook/blob/main/notebooks/cookbook/Saving%20Systems%20to%20XML%20Files.ipynb) what you mean?

rhoslynroebuck replied:
I think Harry means the NanoVer OpenMM tutorial about polyanalanine: https://github.com/IRL2/nanover-protocol/blob/main/examples/openmm/openmm_polyalanine.ipynb

Mohamed-Dhouioui replied:
resolved in commit [e43e5bf](https://github.com/IRL2/nanover-docs/pull/115/commits/e43e5bf8a8b24a180e5f9e6fa04ecfc6947ae248)

hjstroud reviewed (commented):
This is looking really good @Mohamed-Dhouioui. I have left a couple more comments, one of them mentioning the PR about upcoming changes to the xml serializer/deserializer to enable it to save/load the xml state (see IRL2/nanover-protocol#285) which will require minor changes to accommodate when it is completed and merged into main.

I think this page is looking much better, thanks for moving it to the OpenMM specific page, it works nicely there. We're almost there!

hjstroud commented on source/tutorials/openmm.rst:
```diff
@@ -17,4 +23,129 @@ folder of the GitHub repository. It contains:
+Openmm XML Files
+==============================================
+
+The XML format described here is used specifically for manipulating NanoVer simulations from OpenMM serialized XML files. It consists of a root element ``<OpenMMSimulation>`` containing three main components:
```
```suggestion
The XML format described here is used specifically for saving and loading NanoVer OpenMM simulations. It consists of a root element ``<OpenMMSimulation>`` containing three main components:
```
The suggested change here reflects the fact that the xml file also contains non-OpenMM components from the PDB file.

rhoslynroebuck commented on source/tutorials/openmm.rst:
```diff
@@ -2,6 +2,12 @@
+    :depth: 2
+    :local:
+
+Openmm Tutorials
```
I've just noticed that "OpenMM" is written "Openmm" in a few places on this page. We should check and make sure that we consistently write "OpenMM".
```suggestion
OpenMM Tutorials
```

rhoslynroebuck commented:
So just one last thing, you changed the heading from `XML Files` to `OpenMM XML Files`, but then refer to these files as `NanoVer OpenMM XML files" in the rest of the page. Would it be best to change the heading to "NanoVer OpenMM XML files" as well to keep the terminology consistent?

Mohamed-Dhouioui commented:
> Just one last thing, you have changed the heading from `XML Files` to `OpenMM XML Files`, but then refer to these files as `NanoVer OpenMM XML files" in the rest of the page. Would it be best to change the heading to "NanoVer OpenMM XML files" as well to keep the terminology consistent?

I renamed the heading accordingly and added a link to the polyanalanine tutorial notebook.

hjstroud reviewed (approved):
I have added a couple of final touches to the page, including correcting the description of the `platform_name` optional argument and changing Nanover to NanoVer in a title. I think this is now ready to be pushed to main!
